src/main/java/uk/gov/hmcts/reform/logging/filters/RequestIdsSettingFilter.java

```diff
@@ -35,38 +35,52 @@ public RequestIdsSettingFilter(Supplier<String> requestIdGenerator) {
      * {@inheritDoc}.
      */
     @Override
-    public void init(FilterConfig filterConfig) throws ServletException {
+    public void init(FilterConfig filterConfig) {
+        // Nothing to do
     }
 
     /**
      * {@inheritDoc}.
      */
     @Override
-    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
-        throws IOException, ServletException {
+    public void doFilter(
+        ServletRequest request,
+        ServletResponse response,
+        FilterChain chain
+    ) throws IOException, ServletException {
         try {
             HttpServletRequest httpServletRequest = (HttpServletRequest) request;
-            String rootRequestId = httpServletRequest.getHeader(HttpHeaders.ROOT_REQUEST_ID);
-
             MdcFields.setSessionId(getSessionId(httpServletRequest));
-            MdcFields.setRequestId(requestIdGenerator.get());
-            MdcFields.setRootRequestId(rootRequestId == null ? MdcFields.getRequestId() : rootRequestId);
-            MdcFields.setOriginRequestId(httpServletRequest.getHeader(HttpHeaders.ORIGIN_REQUEST_ID));
+            setRequestTracingHeaders(httpServletRequest);
 
             chain.doFilter(request, response);
         } finally {
             MdcFields.removeAll();
         }
     }
 
+    private void setRequestTracingHeaders(HttpServletRequest httpServletRequest) {
+        String requestId = httpServletRequest.getHeader(HttpHeaders.REQUEST_ID);
+        if (isPresent(requestId)) {
+            MdcFields.setRequestId(requestId);
+            String rootRequestId = httpServletRequest.getHeader(HttpHeaders.ROOT_REQUEST_ID);
+            MdcFields.setRootRequestId(isPresent(rootRequestId) ? rootRequestId : requestId);
+            MdcFields.setOriginRequestId(httpServletRequest.getHeader(HttpHeaders.ORIGIN_REQUEST_ID));
+        } else {
+            MdcFields.setRequestId(requestIdGenerator.get());
+            MdcFields.setRootRequestId(MdcFields.getRequestId());
+        }
+    }
+
     private String getSessionId(HttpServletRequest httpServletRequest) {
         HttpSession session = httpServletRequest.getSession(false);
         return session == null ? null : session.getId();
     }
 
-    /**
-     * {@inheritDoc}.
-     */
+    private boolean isPresent(String string) {
+        return string != null && !string.isEmpty();
+    }
```

Comment:
Is Guava or Apache Commons imported? if so, perhaps just `Strings.isNullOrEmpty` (or whatever it is ;) ) could be used?

Reply:
It's not, I looked for it :( Do we want to bring it in just for this?

Reply:
in that case it's perfectly fine